### ISSUE TITLE
Bug: tensors built by tf.keras layers cannot use numpy() to obtain the array. 

### ISSUE THREAD

**Issue by Tenyn** [tensorflow/tensorflow#34787]
Bug: tensors built by tf.keras layers cannot use numpy() to obtain the array. So can i get the value of such tensors??

**Comment by ravikyram** [comment]
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?

Request you to provide simple standalone code to reproduce the issue in our environment. If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!

**Comment by Tenyn** [comment]
Operating system is windows 10, and TensorFlow version s 2.0.0. 
I want to obtain the value of my custom layer, and fail by using numpy() which raises 'Tensor' object has no attribute 'numpy' error. 
Eager execution is used automatically. 

The code is here, using an example in the tensorflow tutorials.

`import tensorflow as tf
from tensorflow.keras import *

print("TensorFlow version: {}".format(tf.__version__))
print("Eager execution: {}".format(tf.executing_eagerly()))

# dataset
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

# build model
import tensorflow.keras.backend as K

class MyDenseLayer(tf.keras.layers.Layer):
    def __init__(self, num_outputs):
        super(MyDenseLayer, self).__init__()
        self.num_outputs = num_outputs
        self.a = None
  
    def build(self, input_shape):
        self.kernel = self.add_weight("kernel",
                                        shape=[int(input_shape[-1]), self.num_outputs])
  
    def call(self, inputs):
        self.a = K.dot(inputs, self.kernel)
        return K.dot(inputs, self.kernel)

layer1 = MyDenseLayer(64)
input_tensor = Input(shape=(28, 28))
x = layers.Flatten()(input_tensor)
x = layer1(x)
output_tensor = layers.Dense(10, activation='softmax')(x)

model = Model(input_tensor,  output_tensor)

model.summary()

# loss function and optimizer
adam = tf.keras.optimizers.Adam()
model.compile(optimizer=adam,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

for epoch in range(2):
    model.fit(x_train, y_train, batch_size=32, epochs=1)

layer = model.layers[2]
print('--------------------------------------------------')
print(layer.a)

layer.a.numpy()`

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0 ,2.1.0-dev20191203 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/de006b8bfcd8c04b89f483a0eaececac/untitled446.ipynb). Thanks!

**Comment by Athul8raj** [comment]
Even though the boolean flag "tf.executing_eagerly()" is True, while using the Keras model you are essentially creating a graph(DAGs) of layers. So in the flow through the graph, tensors created are of the type "tensorflow.python.framework.ops.Tensor" and not "tensorflow.python.framework.ops.EagerTensor". 
Hence the error "'Tensor' object has no attribute 'numpy' error.".

**Comment by Mike575** [comment]
> Even though the boolean flag "tf.executing_eagerly()" is True, while using the Keras model you are essentially creating a graph(DAGs) of layers. So in the flow through the graph, tensors created are of the type "tensorflow.python.framework.ops.Tensor" and not "tensorflow.python.framework.ops.EagerTensor".
> Hence the error "'Tensor' object has no attribute 'numpy' error.".

How to change tensorflow.python.framework.ops.Tensor into numpy?

**Comment by Athul8raj** [comment]
You have to explicitly declare as EagerMode to run it eagerly
tf.config.experimental_run_functions_eagerly(True).

Please note that there will be significant performance hits when this is set to true. This is suggested only in cases like debugging etc.
Even after decorating the __call__ method of  MyDenseLayer with tf.function. there is only slight performance gain.
Please find the gist here https://colab.research.google.com/drive/16hPlAHon5Lwr2uy1Nl3OSAOiBXLqSe4A

**Comment by omalleyt12** [comment]
@Tenyn thanks for the issue!

As @Athul8raj mentioned, this is expected behavior for Tensor created inside a Functional Model. These Tensors are created during the execution of a `tf.function`, and as such are Symbolic Tensors, which cannot be converted to a value

You could assign `layer.a` to a `Variable` if you need to inspect it's value after execution, the `Variable`s have values that live on after `tf.function` execution. Something like this should work:

```python
class MyDenseLayer(tf.keras.layers.Layer):
    def __init__(self, num_outputs):
        super(MyDenseLayer, self).__init__()
        self.num_outputs = num_outputs
        self.a = tf.Variable([0]*num_outputs, dtype=tf.dtypes.float32)
  
    def build(self, input_shape):
        self.kernel = self.add_weight("kernel",
                                        shape=[int(input_shape[-1]), self.num_outputs])
  
    def call(self, inputs):
        self.a.assign(K.dot(inputs, self.kernel))
        return K.dot(inputs, self.kernel)

layer1 = MyDenseLayer(64)
input_tensor = Input(shape=(28, 28))
x = layers.Flatten()(input_tensor)
x = layer1(x)
output_tensor = layers.Dense(10, activation='softmax')(x)

model = Model(input_tensor,  output_tensor)

model.summary()

adam = tf.keras.optimizers.Adam()
model.compile(optimizer=adam,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

for epoch in range(2):
    model.fit(x_train, y_train, batch_size=32, epochs=1)

layer = model.layers[2]
print('--------------------------------------------------')
print(layer.a)

layer.a.numpy()
```
